javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -266,46 +283,47 @@ private void setCurrentArray() {
     BitMappedTrie<T> filter(Predicate<? super T> predicate) {
         Objects.requireNonNull(predicate, "predicate is null");
 
-        final Object[] results = new Object[length()];
+        final Object results = type.newInstance(length());
         final int length = this.<T> visit((index, leaf, start, end) -> {
             for (int i = start; i < end; i++) {
-                final T value = leaf[i];
+                final T value = type.getAt(leaf, i);
                 if (predicate.test(value)) {
-                    results[index++] = value;
+                    type.setAt(results, index++, value);
                 }
             }
             return index;
         });
 
         return (length() == length)
                ? this
-               : BitMappedTrie.ofAll(copyRange(results, 0, length));
+               : BitMappedTrie.ofAll(type.copyRange(results, 0, length));
     }
 
     <U> BitMappedTrie<U> map(Function<? super T, ? extends U> mapper) {
         Objects.requireNonNull(mapper, "mapper is null");
 
-        final Object[] results = newInstance(length);
+        final Object results = obj().newInstance(length);
         this.<T> visit((index, leaf, start, end) -> {
             for (int i = start; i < end; i++) {
-                results[index++] = mapper.apply(leaf[i]);
+                obj().setAt(results, index++, mapper.apply(type.getAt(leaf, i)));
             }
             return index;
         });
 
         return BitMappedTrie.ofAll(results);
     }
 
-    int length() { return length; }
+    int length()                         { return length; }
+    private boolean cannotAdd(T element) { return type != obj() && ((element == null) || (type.type() != asPrimitive(element.getClass()))); }
```

Comment:
We don't need to check the types. Because we already use `T` we only need to check if we want to insert `null` into a primitive array.

```java
private boolean addNullToPrimitiveArray(T element) { return !arrayType.isPrimitive() && element == null; }
```

I don't think we need an additional check if the element is an Object and we want to insert it into an primitive array. If the element is not null it should be unboxed, right?

If it is not unboxed we have to:

```java
private boolean addNullOrObjectToPrimitiveArray(T element) {
    return arrayType.isPrimitive() && (element == null || !arrayType.isPrimitive());
}
```